### ISSUE TITLE
[Spree Upgrade][FrontOffice] Fix shopping/variant_overrides_spec (1 broken test)

### ISSUE THREAD

**Issue by luisramos0** [openfoodfoundation/openfoodnetwork#3118]
Fix spec/features/consumer/shopping/variant_overrides_spec

these errors will have to fixed after or as part of #2559
```
 1) shopping with variant overrides defined creating orders subtracts stock from the override
     Failure/Error:
       expect do
         complete_checkout
       end.to change { v4.reload.count_on_hand }.by(0)
     
       expected `v4.reload.count_on_hand` to have changed by 0, but was changed by -2
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:126:in `block (4 levels) in <top (required)>'
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:125:in `block (3 levels) in <top (required)>'

  2) shopping with variant overrides defined creating orders subtracts stock from stock-overridden on_demand variants
     Failure/Error:
       expect do
         complete_checkout
       end.to change { v6.reload.count_on_hand }.by(0)
     
       expected `v6.reload.count_on_hand` to have changed by 0, but was changed by -2
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:137:in `block (4 levels) in <top (required)>'
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:136:in `block (3 levels) in <top (required)>'

  3) shopping with variant overrides defined creating orders does not show out of stock flags on order confirmation page
     Failure/Error: sleep(0.1) until value = yield
     
     Timeout::Error:
       execution expired
     # ./spec/support/request/web_helper.rb:151:in `sleep'
     # ./spec/support/request/web_helper.rb:151:in `block in wait_until'
     # ./spec/support/request/web_helper.rb:150:in `wait_until'
     # ./spec/support/request/web_helper.rb:157:in `wait_until_enabled'
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:201:in `click_checkout'
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:155:in `block (3 levels) in <top (required)>'
```

**Comment by luisramos0** [comment]
After #2559 one remaining error:

```ruby
 3) shopping with variant overrides defined creating orders does not show out of stock flags on order confirmation page
     Failure/Error: sleep(0.1) until value = yield
     
     Timeout::Error:
       execution expired
     # ./spec/support/request/web_helper.rb:151:in `sleep'
     # ./spec/support/request/web_helper.rb:151:in `block in wait_until'
     # ./spec/support/request/web_helper.rb:150:in `wait_until'
     # ./spec/support/request/web_helper.rb:157:in `wait_until_enabled'
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:201:in `click_checkout'
     # ./spec/features/consumer/shopping/variant_overrides_spec.rb:155:in `block (3 levels) in <top (required)>'
```

**Comment by luisramos0** [comment]
This test was introduced here:
https://github.com/openfoodfoundation/openfoodnetwork/commit/7a5c56cbcabcd78901a364849e9ff945b32916f6

I am having some trouble understanding what's going on here:
- is this a problem with overrides stock management in the cart_controller populate or stock_levels part that leaves the cart dirty on the frontend and blocks the test? or is this a problem with the out of stock alert that should not be shown?
- also, why is the "out of stock" message not be shown for overrides?

any takers? @sauloperez @mkllnk

**Comment by mkllnk** [comment]
I think I know what's going on. We make the variant out of stock, but the variant override still has stock. Since the override is holding the truth, we should be able to check out and we shouldn't see any mention of "Out of stock".

I'm not sure why the checkout fails here. It could be a problem with overrides not being applied. There may be another error making the cart invalid. That's what we need to fix.

**Comment by luisramos0** [comment]
ah, how did I miss that? :-)
thanks a lot @mkllnk 

Ill see what's going on now. I find it weird that there's no scoping done  on stock_levels.on_hand here...
https://github.com/openfoodfoundation/openfoodnetwork/blob/046ba14da4111176270d8e9ddaf8164b9dc59bf7/app/controllers/cart_controller.rb#L33-L46

**Comment by sauloperez** [comment]
I'm so looking forward to end the variant overrides madness...

**Comment by luisramos0** [comment]
So, the current error is related to a new s5 bug in master: #3222 

The fix to #3222 would actually fix this spec BUT this is only getting to the broken code because there's something wrong when adding the variant to the cart here (the line_items are created in this method but after reload order.line_items is empty):
https://github.com/openfoodfoundation/openfoodnetwork/blob/24ca66bd71c13264fa6125f9368f49f73dde349d/app/models/spree/order_decorator.rb#L157-L161
I still don't understand why this is happening.
